src/CofreeBot/Bot.hs

```diff
@@ -47,6 +53,48 @@ newtype Bot m s i o = Bot { runBot :: s -> i -> m (o, s) }
     (Functor, Applicative, Monad, MonadState s, MonadReader i, MonadIO)
   via StateT s (ReaderT i m)
 
+newtype Behavior m i o = Behavior { runBehavior :: i -> m (o, (Behavior m i o)) }
+
+instance Functor m => Functor (Behavior m i)
+  where
+  fmap f (Behavior b) = Behavior $ fmap (fmap (bimap f (fmap f))) b
+
+instance Functor m => Profunctor (Behavior m)
+  where
+  dimap f g (Behavior b) = Behavior $ dimap f (fmap (bimap g (dimap f g))) b
+
+instance Applicative m => Choice (Behavior m)
+  where
+  left' (Behavior b) = Behavior $ either (fmap (bimap Left left') . b) (pure . (, left' (Behavior b)) . Right)
+
+instance Functor m => Strong (Behavior m)
+  where
+  first' (Behavior b) = Behavior $ \(a, c) -> fmap (bimap (, c) first') (b a)
+
+instance Monad m => Traversing (Behavior m)
+  where
+  -- TODO: write wander instead for efficiency
+
+  traverse' b = Behavior $ \is ->
+    fmap (uncurry (,) . fmap traverse')
+    $ flip runStateT b
+    $ traverse (\i -> StateT $ \(Behavior b') -> fmap (\(responses, nextState) -> (responses, nextState)) $ b' i) is
+
+fixBot :: Functor m => Bot m s i o -> s -> Behavior m i o
```

Comment:
imo this would be more satisfying as `Behavior m (i, s) (o, s) -> Behavior m i o`, at which point it's just costrength